apio-consumer/src/main/kotlin/com/liferay/apio/consumer/ApioConsumer.kt

```diff
@@ -0,0 +1,345 @@
+		.build()
+
+private fun parse(
+	response: Response): Result<Thing, Exception> =
+	response.body()?.let {
+		val result = parse(it.string())
+
+		result?.let {
+			val (thing, embeddedThings) = it
+			val nodes = embeddedThings.map { (id, embeddedThing) ->
+				val previousThing = graph[id]?.value
+
+				val newThing = embeddedThing?.merge(previousThing) ?: previousThing
+
+				id to Node(id, newThing)
+			}
+
+			graph.put(thing.id, Node(thing.id, thing))
+			graph.putAll(nodes)
+
+            Result.of(thing)
+        }
+    } ?: Result.of(null, { ApioException("Not Found") })
+
+class Node(val id: String, var value: Thing? = null)
+
+var graph: MutableMap<String, Node> = mutableMapOf()
+
+var credential: String = Credentials.basic("test@liferay.com", "test")
+
+fun parse(json: String): Pair<Thing, Map<String, Thing?>>? {
+	val mapType = TypeToken.getParameterized(Map::class.java, String::class.java, Any::class.java).type
+
+	val jsonObject = Gson().fromJson<Map<String, Any>>(json, mapType)
+
+	return flatten(jsonObject, null)
+}
+
+private fun flatten(jsonObject: Map<String, Any>, parentContext: Context?): Pair<Thing, Map<String, Thing?>>? {
+
+	if (!jsonObject.containsKey("statusCode")) {
+		val id = jsonObject["@id"] as String
+
+		val types = parseType(jsonObject["@type"])
+
+		val context = contextFrom(jsonObject["@context"] as? List<Any>, parentContext)
+
+		val (attributes, things) = foldTree(jsonObject, context)
+
+		val operations = parseOperations(jsonObject)
+
+		val thing = Thing(id, types, attributes, operations =  operations)
+
+		return thing to things
+	}
+	return null
+}
+
+private fun parseOperations(jsonObject: Map<String, Any>): MutableMap<String, Operation> {
+	val operationsJson = jsonObject["operation"] as? List<Map<String, Any>> ?: listOf()
+
+	return operationsJson.map {
+		val id = it["@id"] as String
+		val target = it["target"] as String
+		val method = it["method"] as String
+		val expects = it["expects"] as? String
+		val type = parseType(it["@type"])
+
+		val form = expects?.let { OperationForm(it) }
+
+		id to Operation(id, target, type, method, form)
+	}.toMap().toMutableMap()
+}
+
+private fun parseType(type: Any?): List<String> {
+	return (type as? String)?.let { listOf(it) } ?:
+	type as? List<String> ?: listOf()
+}
+
+
+private fun foldTree(jsonObject: Map<String, Any>,
+	context: Context?): Pair<MutableMap<String, Any>, MutableMap<String, Thing?>> {
+	return jsonObject
+		.filter { it.key !in listOf("@id", "@type", "@context") }
+		.entries
+		.fold(mutableMapOf<String, Any>() to mutableMapOf(), foldEntry(context))
+}
+
+
+typealias FoldedAttributes = Pair<MutableMap<String, Any>, MutableMap<String, Thing?>>
+
+private fun foldEntry(context: Context?) = { acc: FoldedAttributes, entry: Entry<String, Any> ->
+	val (attributes, things) = acc
+
+	val key = entry.key
+	val value = entry.value
+
+	when {
+		value is Map<*, *> -> (value as? Map<String, Any>)?.apply {
```

Comment:
we should extract functions here to know what we are checking

Reply:
I extracted in two functions, the PR is now updated.